### ISSUE TITLE
Template polymorphism breaks Prop/Set separation

### ISSUE THREAD

**Issue by SkySkimmer** [coq/coq#9294]
ie we can get a Prop equivalent to any Set, in this example equivalent to `nat`.
~~~coq
Inductive Foo@{i} (A:Type@{i}) : Type := foo : (nat:Type@{i}) -> Foo A.
Arguments foo {_} _.

Check Foo True : Prop. (* should fail *)

Definition bar : Foo True -> nat := fun '(foo x) => x.

Definition foo_bar n : foo (bar n) = n.
Proof. destruct n;reflexivity. Qed.

Definition bar_foo n : bar (foo n) = n.
Proof. reflexivity. Qed.
~~~

I think we can't actually get a proof of false from this without the kind of axiom which breaks impredicative set but still marking as critical.

The root cause is that we typecheck the inductive with universes considered >= Set but then template polymorphism allows instantiating with Prop.

Not sure if there's a way to trigger this with neither `@{i}` or explicit `#[universes(template)]` syntax (if I don't use the `@{i}` universes unify differently) but plugins could probably do it since a long time ago (maybe since universe polymorphism, maybe before, I don't know how universes worked before).

**Comment by ppedrot** [comment]
> I think we can't actually get a proof of false from this without the kind of axiom which breaks impredicative set but still marking as critical.

That still means that currently Coq is inconsistent with classical logic...

**Comment by ppedrot** [comment]
> I think we can't actually get a proof of false from this without the kind of axiom which breaks impredicative set but still marking as critical.

I wouldn't be that optimistic.

```
Inductive Foo@{i} (A:Type@{i}) : Type := foo : (Set:Type@{i}) -> Foo A.
Arguments foo {_} _.

Definition bar : Foo True -> Set := fun '(foo x) => x.

Definition foo_bar (n : Foo True) : foo (bar n) = n.
Proof. destruct n;reflexivity. Qed.

Definition bar_foo (n : Set) : bar (foo n) = n.
Proof. reflexivity. Qed.

Require Import Hurkens.

Inductive box (A : Set) : Prop := Box : A -> box A.

Definition Paradox : False.
Proof.
unshelve refine (
  NoRetractFromSmallPropositionToProp.paradox
  (Foo True)
  (fun A => foo A)
  (fun A => box (bar A))
  _
  _
  False
).
+ cbn; intros A [x]; exact x.
+ cbn; intros A x; refine (Box _ x).
Qed.
```

**Comment by SkySkimmer** [comment]
Indeed we actually typecheck with the strict `i > Set`

**Comment by maximedenes** [comment]
So when was the bug introduced?

**Comment by SkySkimmer** [comment]
Still around universe polymorphism I think
Or maybe around 4838a3a3c25cc9f7583dd62e4585460aca8ee961

**Comment by SkySkimmer** [comment]
Or before depending on whatever was done with universes before polymorphism was introduced.

**Comment by herbelin** [comment]
> plugins could probably do it since a long time ago (maybe since universe polymorphism, maybe before, I don't know how universes worked before).

This is probably right. In e.g. 8.4, before explicit universe polymorphism, even if this was not intended, a plugin could force a sharing of universes as done in the example above. Then, the constraint `i>=Set` would not be taken into account in computing the sort of the inductive type (function `Indtypes.typecheck_inductive`). It would be inferred to be `Type@{i}` hence put in `Prop` when applied to `True`.

@SkySkimmer: are you going to do a fix?

**Comment by maximedenes** [comment]
So guys, any hope for a fix?

**Comment by SkySkimmer** [comment]
>Indeed we actually typecheck with the strict i > Set

I was wrong, we actually get a context `u |= u > Set`, however we're not checking when instantiating to Set.

**Comment by mattam82** [comment]
So, IIUC, we allow i = Set here implicitely while when typechecking we had i > Set. We could restrict template poly to check that the substitution it builds is allowed, but that would disallow lowering to Prop in most cases, right? E.g for prod True True, we’d have to check i = Prop while i > Set in general. The only sound solution I see is to support polymorphism on Prop (any universe is just >= Prop by default, can be heightened to Set for compatibility with HoTT) and simulate template with full polymorphism.

**Comment by SkySkimmer** [comment]
By the way here's an evolved possibly trickier to handle example:
~~~coq
Universe i. Constraint Set < i.
Inductive foo@{|} (A:Type@{i}) : Type := C : (Set:Type@{i}) -> foo A.
Check foo nat : Set.
~~~

**Comment by mattam82** [comment]
I guess the same reasoning applies, with full polymorphism we wouldn't allow the lowering of the return type. We could try verifying in type_of_inductive_gen that when we instantiate A's universe i by Set, this results in a consistent definition (which would fail here), but as noted above, without allowing some global universes to be instantiated by Prop this would make template-polymorphism ineffective. 
A similar idea would be to make Template Polymorphic inductives return types contain the lower bounds on their parameter universes (e.g. `Foo` would have type `Type@{i,Set+1}` here) but again this would prevent template polymorphism to fire for `prod : Type@{i} -> Type@{i} -> Type@{i,Set+1})`, on `prod nat nat`. @herbelin @SkySkimmer do you see any other way to fix this than:
1 - typechecking allowing fresh levels to be just >= Prop
2 - allowing polymorphic constants to take Prop in instances
3 - emulating template-polymorphic types using fully-polymorphic ones

**Comment by SkySkimmer** [comment]
I feel like trying to fix it another way would just lead to partially reimplementing full polymorphism.

**Comment by gares** [comment]
I'm very naive here, but since day 0 of universe polymorphism I was hoping that 3 would be easy to do. I'm happy this seems doable today.

Just to be sure I understand where we are going: can we say that the argument of `Constr.Sort` is the thing we quantify on (when we are universe polymorphic)? (today is the argument of `fun x => Constr.(Sort (Type x))` if I'm not mistaken). So `Prop` would just be seen as named universe with extra, ad-hoc, properties.

**Comment by maximedenes** [comment]
> So `Prop` would just be seen as named universe with extra, ad-hoc, properties.

Sure, but the difficulty is that these extra powers come with extra restrictions too. Like it has to be at the bottom of the  universe hierarchy, to say things in a simplified way.

So the question is how you enforce these extra restrictions, when you don't necessarily now yet that an abstract universe will be instantiated to `Prop`.

It is surely possible, but we should certainly be rigorous and try to get some correctness proofs on paper at least.

**Comment by SkySkimmer** [comment]
With SProp we have issues due to non-cumulativity, consider
~~~coq
Definition foo@{i j|i <= j} : Type@{i} -> Type@{j} := fun x => x.
Definition bar@{i j|i < j} : Type@{j} := Type@{i}.
Definition baz@{i j|i < j} := foo@{i j}.
~~~
We should not be allowed to do `foo@{SProp u}`, `bar@{SProp u}` may be OK (with `Set < u`), but `baz@{SProp u}` is not OK even though it has the same constraints as `bar`.
ie if we have a strict constraint we don't know if the implicit nonstrict constraint is needed (in which case instantiating with SProp is not allowed).
In other words if we get a constraint between SProp and some other universe, we need to make the other universe equal to SProp. If the constraint is strict we can never satisfy it.
It might still be possible to instantiate with SProp when there are no constraints but I don't know how useful that would be.

With Prop you need to have `i <= Set` imply `i = Prop \/ i = Set` and `i < Set` forbidden.

**Comment by SkySkimmer** [comment]
>today is the argument of fun x => Constr.(Sort (Type x)) if I'm not mistaken

Not really, we can already instantiate with Set.

**Comment by mattam82** [comment]
Indeed, one issue with Prop is that one could like to maintain that there is no universe strictly between Prop and Set, although I don't think that matters for consistency: if Prop < i < Set then we still have a well-founded predicative hierarchy above Prop (assuming i is treated predicatively). Ensuring that everything is >= Prop is easy, we just have to introduce such a constraint with each universe (instead of what is done currently with >=/> Set). I'd leave SProp alone for now, but indeed @gares, this would go towards removing Prop/Set in Sorts and only use their representations in Type u (type0/type0_m_univ precisely).

**Comment by gares** [comment]
> how you enforce these extra restrictions

I think that *instantiation* (and I guess also the addition of an = constraint) should then be allowed to add constraints. I don't know today how instantiation is typed, but for sure unification can.

More concretely if you want to say that `i = Prop` (by instantiation or unification) then you have to check nothing `_ < i` exists and that if `i` was eliminated to inhabit `j` then one should enforce `j = Prop` as well (or alternatively say that in order to eliminate an unknown `i` for `j` one has that `Prop < i`, and hence prevent `i` to be instantiated/unified with `Prop` later on, but this may be a problem to implement template-polymorphic-induction-schemas).

**Comment by gares** [comment]
Anyway, I think we should make good use of the upcoming working group and discuss the design of universes all together.

**Comment by herbelin** [comment]
For the record, three other variants of the problem indirectly pointed out by @mattam82 in PR #9918 (section `Foo` of file `test-suite/failure/Template.v`), either using ltac tricks to create a sharing between universes:
```
Section Foo.
Context (A : Type).
Definition CastSet (_:Type) := Set : ltac:(let ty := type of A in exact ty).
Inductive Foo := foo : CastSet A -> Foo.
End Foo.
Check Foo True : Prop.
```
```
Inductive Foo A (S:= fun _ => Set : ltac:(let ty := type of A in exact ty)
  := foo : S A -> Foo A.
Check Foo True : Prop.
```
```
Polymorphic Definition CastSet (A:Type) := Set : ltac:(let ty := type of A in exact ty).
Inductive Foo A := foo : CastSet A -> Foo A.
Check Foo True : Prop.
```
All wrongly succeed by not taking into account the lower constraints on the level of the type of `A`.

**Comment by herbelin** [comment]
Ah, great. Then here is a new step of simplification for 8.4 and later (no need for a phantom type actually, and the outer `match` can be removed):
```
Inductive Foo A := foo : (Set : match _ with T => match A : T with _ => T end end) -> Foo A.
Check foo _ bool : Foo True : Prop.
```
I cannot make it in 8.2 and 8.3 though: unification fails on the same example.

In any case, this means that the implicit assumption that universe levels in template polymorphism occurred only once at the time of declaration of the parameter was too crude (unification, pattern-matching compilation, Ltac, ... can introduce sharing).

**Comment by herbelin** [comment]
> > In any case, this means that the implicit assumption that universe levels in template polymorphism occurred only once at the time of declaration of the parameter was too crude (unification, pattern-matching compilation, Ltac, ... can introduce sharing).

> It seems plausible to me that this assumption was true at the time template polymorphism was introduced, but was not recorded anywhere (or was forgotten) and was violated as the system became more powerful.

I unfortunately don't remember all the details from that time. The basic use case was clear and the possibility of border cases breaking the basic assumptions apparently did not come to the mind. Bruno B. discovered in 2011 that versions 8.1 to 8.3 had a [critical bug](https://github.com/coq/coq/blob/f9e8058faac1831082b27d9f5ba3c45e340b4c7e/dev/doc/critical-bugs#L100-L110) involving universe aliasing, so, at least this case had not been imagined. At that time, it seems that I had been too confident that template polymorphism could primarily be seen as a macro avoiding code duplication, neglicting the possibility of unexpected border cases. But note that even after the 2011 aliasing problem, it was not thought that a universe variable in the type of a parameter could generate a constraint from below.

Anyway, full universe polymorphism now allows us to emulate template polymorphism using concepts supporting constraints from below (explicit quantification over variables and over a graph of constraints).
